docs/rfcs/0004-rfc-connectiomanager.md

```diff
@@ -0,0 +1,74 @@
+
+## Proposed Solution
+
+The Connection Diagnostics will be responsible for managing all probes. These probes are stored as distinct states within an enumeration class, `ConnectionDiagnostics::State`. The Connection Diagnostics will sequentially cycle through each of these states in a systematic order, only proceeding to the next state if the current probe is successful. 
+
+In the event of a probe failure, the Connection Diagnostics will emit a signal to the frontend. This will trigger a modal to be displayed to the user, providing an explanation of the issue and suggesting potential solutions. Upon encountering a failed probe, the progression through the remaining probes is immediately halted and the Connection Diagnostics will enter the Idle state.
+
+The Connection Health, an existing component of our codebase, is tasked with sending periodic pings and determining the stability of the VPN based on the responses received (see `ConnectionHealth::healthCheckup()`). It classifies the VPN stability as `Stable`, `Unstable` (some missing responses), or `NoSignal` (no responses received). When the VPN experiences issues with network connectivity or server availability, it naturally transitions into the No Signal state and, ideally, recovers once the issue is resolved. We can leverage this logic to recover from a probe failure within the Connection Diagnostics.
+
+> For specific details about how these failures are handled and communicated with the user, please refer to this [Figma document](https://www.figma.com/file/GZKigVZbnCisMJNU3Hc9xM/Connection-error-messaging?type=design&node-id=26-215&mode=design&t=kfYnUKuXQ9nsGDRR-0).
+
+## Implementation Plan
+
+```c++
+enum ConnectionDiagnosticsStates {
+  StateProbingInternet,
+  StateCheckingFirewall,
+  StateProbingServerLocation,
+  StateCheckingSubscription,
+  StateIdle,
+}
+```
+
+### Upon VPN Activation (`StateOff` to `StateConnecting`) - Phase 1
+Perform a single cycle going through all the Connection Diagnostics probes.
+- If a probe fails:
+  - Set Controller to `StateOff`
+  - Stop subsequent probes from running
+  - emit a signal to the frontend communicating which probe failed
+  - VPN will remain in `StateOff`
+  - Connection Diagnostics will go into `StateIdle`
+- If all probes pass:
+  - Activate VPN (Controller `StateOn`)
+  - Connection Diagnostics will go into `StateIdle`
+
+### VPN is already active (`StateOn`) - Phase 2
+If the VPN is already active and Connection Health stability goes into no signal, initiate a diagnostics cycle to perform a set of checks to identify the underlying cause.
```

Comment:
`SubscriptionMonitor` and `CaptivePortalDetection` currently listen for the stabilitychanged signal being emitted by health monitoring. This should stop, right, as it’s now going to be handled by connection diagnostic tool?

Reply:
`CaptivePortalDetection` changes are still ongoing but as a result of a separate effort, moving forward we will not check for captive portal after the VPN is activated, this is also why I didn't include Captive portal in the list of probes.

Regarding Subscription Monitor, yes you are correct. It will be moved to Diagnostics

Reply:
Sorry, missed that it wasn't part of this. Sounds good.

Reply:
What is the `CaptivePortalDetection` out of curiosity? Do you know if that effort will move it into this diagnostic class?

Reply:
We no longer check for captive portal upon network changes, only once during activation. Even though we no longer check for that after the VPN is already activated, I think it should still be in the Diagnostics so I'll add it to the states in this doc. 

Does this answer your question?

Reply:
I didn't know this change was coming, thank you! In relation to this doc - does that mean a captive portal check will be part of the monitoring, or not?

Reply:
Nevermind, I just saw your update - sorry. This sounds great, thanks.